Getting input.value for number inputs (type=number) with over 39 characters returns an empty string

Ahmad-S792 opened the pull request:
#### 134163e4e3101c02038c4447672158e4a2fac445
<pre>
Getting input.value for number inputs (type=number) with over 39 characters returns an empty string

Getting input.value for number inputs (type=number) with over 39 characters returns an empty string
<a href="https://bugs.webkit.org/show_bug.cgi?id=249783">https://bugs.webkit.org/show_bug.cgi?id=249783</a>

Reviewed by Ryosuke Niwa.

This patch is to align WebKit with Blink / Chromium, Firefox / Gecko and Web-Specification.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=168513

Number input type maximum value increased from float to double.
Float allowed only max value of 3.402823466e38, it has now been
increased to 1.7976931348623158e+308, by using double.

InvalidStateError exception is removed from
NumberInputType::setValueAsDouble and
NumberInputType::setValueAsDecimal, as HTMLInputType::setValueAsNumber checks if value is infinite.

HTMLParserIdioms has also been updated as the functions are
triggered in NumberInputType functions and has check for max
values, thus updated.

* Source/WebCore/html/NumberInputType.cpp:
(NumberInputType::valueAsDouble):
- Updated Comment and also removed FIXME
(NumberInputType::valueAsDecimal):
- Updated Comment and also removed FIXME
(NumberInputType::createStepRange):
(1) Remove &quot;FIXME&quot;
(2) Use &quot;doubleMax&quot; instead of &quot;floatMax&quot;
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(parseToDecimalForNumberType):
(1) Updated Comment and also removed FIXME
(2) Use &quot;doubleMax&quot; instead of &quot;floatMax&quot;
(3) Update condition to ASSERT
* LayoutTests/fast/forms/number/number-valueasnumber.html: Updated
* LayoutTests/fast/forms/number/number-valueasnumber-expected.txt: Ditto
* LayoutTests/fast/forms/number/number-stepup-stepdown.html: Updated
* LayoutTests/fast/forms/number/number-stepup-stepdown-expected.txt: Ditto
* LayoutTests/fast/forms/number/number-stepup-stepdown-from-renderer.html: Updated
* LayoutTests/fast/forms/number/number-stepup-stepdown-from-renderer-expected.txt: Ditto

Canonical link: <a href="https://commits.webkit.org/258614@main">https://commits.webkit.org/258614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b93dcc4ea42526813d8ab9668d9f35ff2e5839a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111683 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171879 "Build was cancelled. Recent messages:Pull request contains relevant changes; Deleted stale build files; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile WebKit; Reverted pull request changes; Compiled WebKit (cancelled)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12516 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2437 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94691 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109411 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108164 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9582 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92849 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37297 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91450 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24333 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5015 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25756 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2196 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45249 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5921 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6907 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/forms/number/number-stepup-stepdown-expected.txt:
```diff
@@ -1,174 +1,64 @@
 PASS stepDown("1", null, null, 2) is "-1"
 PASS stepDown("-1", null, null, -1) is "0"
-
-Fractional cases
```
Why are we deleting these test cases? Can we rebaseline instead?

Ahmad-S792 replied:
Cool, let me do it. Thanks!

Ahmad-S792 replied:
Fixed.

rniwa commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -165,8 +156,8 @@ StepRange NumberInputType::createStepRange(AnyStepHandling anyStepHandling) cons
     };
-    Decimal minimum = extractBound(minAttr, -floatMax);
-    Decimal maximum = extractBound(maxAttr, floatMax);
+    Decimal minimum = extractBound(minAttr, -doubleMax);
```
Can we use double min here instead of -max?

Ahmad-S792 replied:
It gives build error and it seems that we don't have any "doubleMin" in Decimal.cpp.

rniwa commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -156,9 +155,8 @@ double parseToDoubleForNumberType(StringView string, double fallbackValue)
-    // See HTML5 2.5.4.3 `Real numbers.'
-    if (-std::numeric_limits<float>::max() > value || value > std::numeric_limits<float>::max())
+    // Numbers are considered finite IEEE 754 Double-precision floating point values.
+    if (-std::numeric_limits<double>::max() > value || value > std::numeric_limits<double>::max())
```
I don't think these conditions will ever hold since double can't hold values beyond min/max of double. Delete?

Ahmad-S792 replied:
Done - now running through EWS and let's see. Meanwhile I fixed other test cases so will only get "infinity" related failures.

Ahmad-S792 replied:
@rniwa - With ASSERT, the input tests are crashing in `mac-AS-debug-wk2`.

Ahmad-S792 replied:
Fixed. Thanks for pointing my silly error over Slack. Thanks!

Ahmad-S792 committed:
Getting input.value for number inputs (type=number) with over 39 characters returns an empty string

Getting input.value for number inputs (type=number) with over 39 characters returns an empty string
https://bugs.webkit.org/show_bug.cgi?id=249783

Reviewed by Ryosuke Niwa.

This patch is to align WebKit with Blink / Chromium, Firefox / Gecko and Web-Specification.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=168513

Number input type maximum value increased from float to double.
Float allowed only max value of 3.402823466e38, it has now been
increased to 1.7976931348623158e+308, by using double.

InvalidStateError exception is removed from
NumberInputType::setValueAsDouble and
NumberInputType::setValueAsDecimal, as HTMLInputType::setValueAsNumber checks if value is infinite.

HTMLParserIdioms has also been updated as the functions are
triggered in NumberInputType functions and has check for max
values, thus updated.

* Source/WebCore/html/NumberInputType.cpp:
(NumberInputType::valueAsDouble):
- Updated Comment and also removed FIXME
(NumberInputType::valueAsDecimal):
- Updated Comment and also removed FIXME
(NumberInputType::createStepRange):
(1) Remove "FIXME"
(2) Use "doubleMax" instead of "floatMax"
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(parseToDecimalForNumberType):
(1) Updated Comment and also removed FIXME
(2) Use "doubleMax" instead of "floatMax"
(3) Update condition to ASSERT
* LayoutTests/fast/forms/number/number-valueasnumber.html: Updated
* LayoutTests/fast/forms/number/number-valueasnumber-expected.txt: Ditto
* LayoutTests/fast/forms/number/number-stepup-stepdown.html: Updated
* LayoutTests/fast/forms/number/number-stepup-stepdown-expected.txt: Ditto
* LayoutTests/fast/forms/number/number-stepup-stepdown-from-renderer.html: Updated
* LayoutTests/fast/forms/number/number-stepup-stepdown-from-renderer-expected.txt: Ditto

Canonical link: https://commits.webkit.org/258614@main
